src/hotspot/share/opto/inlinetypenode.cpp

```diff
@@ -258,9 +258,15 @@ bool InlineTypeNode::field_is_null_free(uint index) const {
 }
 
 void InlineTypeNode::make_scalar_in_safepoint(PhaseIterGVN* igvn, Unique_Node_List& worklist, SafePointNode* sfpt) {
-  // Don't scalarize larvals in their own constructor call because the constructor will update them
-  if (is_larval() && sfpt->is_CallJava() && sfpt->as_CallJava()->method() != nullptr && sfpt->as_CallJava()->method()->is_object_constructor() &&
-      sfpt->as_CallJava()->method()->holder()->is_inlinetype() && sfpt->in(TypeFunc::Parms) == this) {
+  // We should not scalarize larvals in their own inline type constructor methods because their fields could still be
+  // updated. If we scalarize now and update the fields, we cannot propagate the update out of the constructor call
+  // because the scalarized fields are local to this method. We need to use the buffer to make the update visible to
+  // the outside (see also CompiledEntrySignature::compute_calling_conventions()).
+  if (is_larval() && sfpt->is_CallJava() && sfpt->as_CallJava()->method() != nullptr &&
+      sfpt->as_CallJava()->method()->is_object_constructor() && bottom_type()->is_inlinetypeptr() &&
+      sfpt->in(TypeFunc::Parms) == this) {
+    // Receiver of larval is always buffered in its constructor call because it was initially created outside the
+    // constructor.
```

Comment:
```suggestion
    // Receiver is always buffered because it's passed as oop, see special case in CompiledEntrySignature::compute_calling_conventions().
```